istio/discovery.go

```diff
@@ -0,0 +1,565 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &models.ControlPlaneConfiguration{
+		IstioMeshConfig: *istioConfigMapInfo,
+	}, nil
+}
+
+// istioConfigMapName guesses the istio configmap name.
+func istioConfigMapName(conf *config.Config, revision string) string {
+	if conf.ExternalServices.Istio.ConfigMapName != "" {
+		return conf.ExternalServices.Istio.ConfigMapName
+	}
+
+	// If the revision is set, we should use the revisioned configmap name
+	// otherwise the hardcoded 'istio' value is used.
+	configMapName := "istio" // As of 1.19 this is hardcoded in the helm charts.
+	if revision != "default" && revision != "" {
+		configMapName = configMapName + "-" + revision
+	}
+
+	return configMapName
+}
+
+// Discovery detects istio infrastructure and configuration across clusters.
+type Discovery struct {
+	conf           *config.Config
+	kialiCache     cache.KialiCache
+	kialiSAClients map[string]kubernetes.ClientInterface
+}
+
+// NewDiscovery initializes a new Discovery.
+func NewDiscovery(kialiSAClients map[string]kubernetes.ClientInterface, cache cache.KialiCache, conf *config.Config) *Discovery {
+	return &Discovery{
+		conf:           conf,
+		kialiCache:     cache,
+		kialiSAClients: kialiSAClients,
+	}
+}
+
+// IsRemoteCluster determines if the cluster has a controlplane or if it's a remote cluster without one.
+// Clusters that do not exist or are not accessible are considered remote clusters.
+func (in *Discovery) IsRemoteCluster(ctx context.Context, cluster string) bool {
+	mesh, err := in.Mesh(ctx)
+	if err != nil {
+		log.Debugf("Unable to get mesh to determine if cluster [%s] is remote. Err: %s", cluster, err)
+		return false
+	}
+
+	// If there's a controlplane for the cluster then it's not a remote cluster.
+	for _, controlPlane := range mesh.ControlPlanes {
+		if controlPlane.Cluster.Name == cluster {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Clusters resolves the Kubernetes clusters that are hosting the mesh. Resolution
+// is done as best-effort using the resources that are present in the cluster.
+// TODO: should this go in kubernetes package?
+func (in *Discovery) Clusters() ([]models.KubeCluster, error) {
+	if clusters := in.kialiCache.GetClusters(); clusters != nil {
+		return clusters, nil
+	}
+
+	// Even if somehow there are no clusters found, which there should always be at least the homecluster,
+	// setting this to an empty slice will prevent us from trying to resolve again.
+	clustersByName := map[string]models.KubeCluster{}
+	for cluster, client := range in.kialiSAClients {
+		info := client.ClusterInfo()
+		meshCluster := models.KubeCluster{
+			// If there's a client for this cluster then it's accessible.
+			Accessible: true,
+			Name:       cluster,
+			SecretName: info.SecretName,
+		}
+		if info.ClientConfig != nil {
+			meshCluster.ApiEndpoint = info.ClientConfig.Host
+		}
+
+		meshCluster.Network = in.resolveNetwork(cluster)
+
+		if cluster == in.conf.KubernetesConfig.ClusterName {
```

Comment:
Kind of and this done when kiali first starts up here: https://github.com/kiali/kiali/blob/5f88cc10d836e7f4053a44f9324ec005382e7da3/kiali.go#L220-L265

Kubernetes has no notion of cluster name. That comes from istio and kiali tries to autodiscover it by reading the `CLUSTER_ID` env var on an istiod deployment in the cluster that kiali is running in. If istio is down when kiali tries to do the autodiscovery then kiali will default it to the istio cluster name which is `Kubernetes`. But ultimately the safest way to set the cluster name is for users to set this manually.

The reason the autodiscovery of cluster name happens outside of the rest of the autodiscovery here is because the kiali service account clients are created in a map with their cluster name and this istio discovery service needs those service account clients so we need the cluster name first before we can use this auto discovery service.